File: Awesome.AI.Source/Awesome.AI/Core/Internals/ModProperties.cs
using Awesome.AI.Source.Awesome.AI.Common;
using Awesome.AI.Variables;
using static Awesome.AI.Variables.Enums;

namespace Awesome.AI.Core.Internals
{
    public class ModProperties
    {
        public class MyModifiers
        {
            public double Mod(TheMind mind, double value, string prop)
            {
                double res = 0.0d;

                if (prop == "base")
                    res = value;
                else
                {
                    switch (prop)
                    {
                        case CONST.prop2_temperament:
                            PATTERN patt = mind.mech.mp.pattern_curr;
                            if (patt == PATTERN.MOODGENERAL) res = mind.calc.Normalize(value, 0.0d, 100.0d, 10.0d, 90.0d);
                            if (patt == PATTERN.MOODGOOD) res = mind.calc.Normalize(value, 0.0d, 100.0d, 60.0d, 90.0d);
                            if (patt == PATTERN.MOODBAD) res = mind.calc.Normalize(value, 0.0d, 100.0d, 10.0d, 40.0d);
                            break;
                        case CONST.prop2_brain: throw new Exception("ModProperties, Mod 1");
                        case CONST.prop3_brain: throw new Exception("ModProperties, Mod 2");
                        case CONST.prop2_comm: throw new Exception("ModProperties, Mod 3");
                        case CONST.prop3_comm: throw new Exception("ModProperties, Mod 4");
                        case CONST.prop4_comm: throw new Exception("ModProperties, Mod 5");
                        default: throw new Exception("ModProperties, Mod 6");
                    }
                }

                return res;                
            }
        }

        public class MyMatrix
        {
            private Dictionary<(string, string), double> _data = new();

            public double this[string key1, string key2]
            {
                get
                {
                    if (_data.TryGetValue((key1, key2), out var value))
                        return value;

                    return 1.0;
                }
                set => _data[(key1, key2)] = value;
            }

            public double Run(double val, string key2)
            {
                double res = 1.0d;

                foreach (var v in _data)
                    res *= this[v.Key.Item1, key2];

                return res * val;
            }
        }

        private MyModifiers Mods { get; set; }
        private MyMatrix Matrix { get; set; }
        public Dictionary<string, double> PropsIn { get; set; }
        public Dictionary<string, double> PropsOut { get; set; }
        
        private TheMind mind;
        private ModProperties() { }
        public ModProperties(TheMind mind, PROPS props)
        {
            this.mind = mind;

            GetProps(props, out Dictionary<string, double> attr, out Dictionary<(string, string), double> matr);

            PropsOut = new Dictionary<string, double>();
            PropsIn = attr;
            Mods = new MyModifiers();
            Matrix = new MyMatrix();

            foreach (var kv in attr)
                PropsOut.Add(kv.Key, 0.0d);

            foreach (var kv in matr)
                Matrix[kv.Key.Item1, kv.Key.Item2] = kv.Value;
        }

        public void Update()
        {
            PROPS props = mind.bot.props;

            GetBase(props, out double _norm);

            foreach (var kv in PropsIn)
            {
                string prop = kv.Key;
                double res = 0.0d;

                res = Mods.Mod(mind, _norm, prop);
                //res = Matrix.Run(res, prop);
                PropsOut[prop] = res;
            }
        }

        private void GetProps(PROPS props, out Dictionary<string, double> attr, out Dictionary<(string, string), double> matr)
        {
            switch (props)
            {
                case PROPS.TEMPERAMENT:
                    //will, mood
                    attr = new Dictionary<string, double> { { "base", double.NaN }, { CONST.prop2_temperament, 2.0 } };

                    matr = new Dictionary<(string, string), double>();
                    matr.Add(("base", CONST.prop1_temperament), 1.0d);
                    break;

                case PROPS.BRAINWAVE:
                    //conflict, readyness
                    attr = new Dictionary<string, double> { { "base", double.NaN }, { CONST.prop2_brain, 2.0 }, { CONST.prop3_brain, 1.2 } };

                    matr = new Dictionary<(string, string), double>();
                    matr.Add(("base", CONST.prop3_brain), 1.0d);
                    matr.Add((CONST.prop2_brain, CONST.prop3_brain), 1.0d);
                    break;

                case PROPS.COMMUNICATION:
                    //opinion, temporality, abstraction
                    attr = new Dictionary<string, double> { { "base", double.NaN }, { CONST.prop2_comm, 2.0 }, { CONST.prop3_comm, 0.5 }, { CONST.prop4_comm, 1.2 } };

                    matr = new Dictionary<(string, string), double>();
                    matr.Add(("base", CONST.prop3_comm), 0.65d);
                    matr.Add((CONST.prop2_comm, CONST.prop3_comm), 0.45d);
                    matr.Add((CONST.prop4_comm, CONST.prop2_comm), 0.35d);

                    break;

                default: throw new NotImplementedException("Properties, GetProps");
            }
        }

        private void GetBase(PROPS props, out double _norm)
        {
            switch (props)
            {
                case PROPS.TEMPERAMENT: _norm = mind.mech.mp.eprops.Will().Norm100VV(mind); break;
                //case PROPS.BRAINWAVE: _norm = mind.mech.mp.eprops.Will().Norm0(mind, 0.0d, 100.0d); break;
                //case PROPS.COMMUNICATION: _norm = mind.mech.mp.eprops.Will().Norm0(mind, 0.0d, 100.0d); break;
                default: throw new NotImplementedException("Properties, GetProps");
            }
        }
    }
}

File: Awesome.AI.Python/Core/Internals/ModProperties.py
from awesome_ai.Common.MyNormalize import Norm100VV
from awesome_ai.Variables.Constants import CONST
from awesome_ai.Variables.Enums import PATTERN, PROPS


class ModProperties:
    class MyModifiers:
        def Mod(self, mind, value, prop):
            if prop == "base": return value
            if prop == CONST.prop2_temperament:
                pattern = mind.mech.mp.pattern_curr
                if pattern == PATTERN.MOODGENERAL: return mind.calc.Normalize(value, 0, 100, 10, 90)
                if pattern == PATTERN.MOODGOOD: return mind.calc.Normalize(value, 0, 100, 60, 90)
                if pattern == PATTERN.MOODBAD: return mind.calc.Normalize(value, 0, 100, 10, 40)
                return 0.0
            messages = {CONST.prop2_brain: "ModProperties, Mod 1", CONST.prop3_brain: "ModProperties, Mod 2",
                        CONST.prop2_comm: "ModProperties, Mod 3", CONST.prop3_comm: "ModProperties, Mod 4",
                        CONST.prop4_comm: "ModProperties, Mod 5"}
            raise RuntimeError(messages.get(prop, "ModProperties, Mod 6"))

    class MyMatrix:
        def __init__(self): self._data = {}
        def __getitem__(self, keys): return self._data.get(tuple(keys), 1.0)
        def __setitem__(self, keys, value): self._data[tuple(keys)] = value
        def Run(self, val, key2):
            result = 1.0
            for key1, _ in self._data: result *= self[key1, key2]
            return result * val

    def __init__(self, mind, props):
        self.mind = mind
        attr, matrix = self.GetProps(props)
        self.PropsOut = {key: 0.0 for key in attr}
        self.PropsIn = attr
        self.Mods, self.Matrix = self.MyModifiers(), self.MyMatrix()
        for keys, value in matrix.items(): self.Matrix[keys] = value

    def Update(self):
        normalized = self.GetBase(self.mind.bot.props)
        for prop in self.PropsIn: self.PropsOut[prop] = self.Mods.Mod(self.mind, normalized, prop)

    def GetProps(self, props):
        if props == PROPS.TEMPERAMENT:
            return ({"base": float("nan"), CONST.prop2_temperament: 2.0}, {("base", CONST.prop1_temperament): 1.0})
        if props == PROPS.BRAINWAVE:
            return ({"base": float("nan"), CONST.prop2_brain: 2.0, CONST.prop3_brain: 1.2},
                    {("base", CONST.prop3_brain): 1.0, (CONST.prop2_brain, CONST.prop3_brain): 1.0})
        if props == PROPS.COMMUNICATION:
            return ({"base": float("nan"), CONST.prop2_comm: 2.0, CONST.prop3_comm: .5, CONST.prop4_comm: 1.2},
                    {("base", CONST.prop3_comm): .65, (CONST.prop2_comm, CONST.prop3_comm): .45,
                     (CONST.prop4_comm, CONST.prop2_comm): .35})
        raise NotImplementedError("Properties, GetProps")

    def GetBase(self, props):
        if props in (PROPS.TEMPERAMENT, PROPS.BRAINWAVE, PROPS.COMMUNICATION):
            return Norm100VV(self.mind.mech.mp.eprops.Will(), self.mind)
        raise NotImplementedError("Properties, GetProps")
